backend/modules/parsers/videoparser.py

```diff
@@ -0,0 +1,162 @@
+
+
+class VideoParser(BaseParser):
+    """
+    VideoParser is a parser class for extracting text from video input.
+    {
+        ".mp4": {
+            "name": "VideoParser",
+            "parameters": {
+                "vlm_model_configuration": {
+                    "name" : "truefoundry/openai-main/gpt-4o-mini"
+                },
+                "audio_model_configuration": {
+                    "name": "faster-whisper/Systran/faster-distil-whisper-large-v3"
+                },
+                "max_chunk_size": 2000,
+                "prompt": "Descibe the given information present in the image - textual/charts/graphs/tables in detail."
+                "total_frames": 15 // total images to be process from video, for testing puposes, if not provided all the images will be processed
+            }
+        }
+    }
+    """
+
+    supported_file_extensions = [
+        ".mp4",
+        ".avi",
+        ".mov",
+        ".flv",
+        ".wmv",
+        ".webm",
+        ".mkv",
+        ".mpg",
+        ".mpeg",
+        ".ogv",
+        ".3gp",
+    ]
+
+    def __init__(
+        self,
+        vlm_model_configuration: ModelConfig,
+        audio_model_configuration: ModelConfig,
+        max_chunk_size: int = 2000,
+        prompt: str = "",
+        total_frames: int = -1,
+        **kwargs,
+    ):
+        """
+        Initializes the VideoParser object.
+        """
+
+        self.vlm_model_configuration = ModelConfig.model_validate(
+            vlm_model_configuration
+        )
+        self.audio_model_configuration = ModelConfig.model_validate(
+            audio_model_configuration
+        )
+        self.max_chunk_size = max_chunk_size
+        self.prompt = prompt
+        self.total_frames = (
+            total_frames  # total images to be process from video, for testing puposes
+        )
+
+        super().__init__(**kwargs)
+
+    def _get_images_from_video(self, video_filepath: str) -> str:
+        """
+        Extract images from video frames,
+        We extract 1 frame every 5 seconds of the video
+        Save all the images in temporary directory
+        """
+        _, file_name = os.path.split(video_filepath)
+        file_name, _ = os.path.splitext(file_name)
+        clip = VideoFileClip(video_filepath)
+        temp_images_dir = tempfile.mkdtemp()
+        clip.write_images_sequence(
+            os.path.join(temp_images_dir, file_name + "_frame%04d.png"),
+            fps=0.2,  # configure this for controlling frame rate.
+        )
+        return temp_images_dir
+
+    def _get_audio_from_video(self, video_filepath: str) -> str:
+        """
+        Extract audio from video
+        """
+        _, file_name = os.path.split(video_filepath)
+        # Save the audio in a temporary directory
+        temp_audio_dir = tempfile.mkdtemp()
+        clip = VideoFileClip(video_filepath)
+        clip.audio.write_audiofile(os.path.join(temp_audio_dir, file_name + ".wav"))
```

Comment:
Is there a async interface for this?

Reply:
Did not find any.
Ref:
- https://moviepy-tburrows13.readthedocs.io/en/improve-docs/ref/VideoClip/VideoFileClip.html
- https://www.llamaindex.ai/blog/multimodal-rag-for-advanced-video-processing-with-llamaindex-lancedb-33be4804822e